runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4788,341 +4800,294 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
-
-#if defined(J9VM_GC_SPARSE_HEAP_ALLOCATION)
-   bool isOffHeapAllocationEnabled = TR::Compiler->om.isOffHeapAllocationEnabled();
-#endif /* J9VM_GC_SPARSE_HEAP_ALLOCATION */
-
-   static char * disableBatchClear = feGetEnv("TR_DisableBatchClear");
-   // If batch clear is disabled, set element's size and mustBeZero fields to 0
-   if (disableBatchClear)
-      {
-      // Zero size arrays are considered "discontiguous" and the "mustBeZero" field of discontiguous arrays must be located where the "size" field of contiguous arrays is.
-      // With this design, we can check the size of an array as if it is contiguous, if the size is not zero then array is contiguous otherwise it is discontiguous.
-      // Check runtime/oti/j9nonbuilder.h for layout details. Simplified layout would look like:
-      // Discontiguous array layout: |  class  | mustBeZero |  size   | ...
-      // Contiguous array layout:    |  class  |    size    | ...
-      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField()/* = "mustBeZero" offset!*/, cg));
-      iComment("Init 1st dim mustBeZero field.");
-      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg));
-      iComment("Init 1st dim size field.");
-   if (TR::Compiler->om.compressObjectReferences())
+      cursor = generateRSInstruction(cg, TR::InstOpCode::SLA, node, dim2SizeReg, trailingZeroes(componentSize), cursor);
+      // Only a positive int32 with no overflow is acceptable. This will guarantee no overflow in multiplication instruction.
+      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRNP, node, inlineAllocFailLabel, cursor);
+      }
+   int32_t headerSize= TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+   bool alignSecondDim = !(OMR::aligned(headerSize, alignmentConstant) && OMR::aligned(componentSize, alignmentConstant));
+   cursor = generateRIInstruction(cg, TR::InstOpCode::AGHI, node, dim2SizeReg, alignSecondDim ? (headerSize + alignmentConstant - 1) : headerSize, cursor);
+   if (alignSecondDim)
+      cursor = generateRILInstruction(cg, TR::InstOpCode::NILF, node, dim2SizeReg, -alignmentConstant, cursor);
+   cursor = generateRRInstruction(cg, TR::InstOpCode::LGR, node, sizeReg, dim2SizeReg, cursor);
+
+   cursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, zeroSecondDimLabel, cursor);
+   // If jumping from zero second dim match, set second dim size to discontiguous. Otherwise this wont change values.
+   cursor = generateRRInstruction(cg, TR::InstOpCode::LR, node, dim2SizeReg, sizeReg, cursor);
+   // Total size = (second dim size * number of second dim arrays) + first dim size.
+   // Max possible value is uint32.max * int32.max which will not overflow.
```

Comment:
Which operand is uint32? I think this comment is bit misleading - Please add more context with the assumption that you are taking - which is max number of bytes would not overflow 64-bit range.